nemo_aligner/algorithms/generation.py

```diff
@@ -0,0 +1,508 @@
+        ):
+            # if you use MegatronPretrainingBatchSampler as the batch_sampler passed to your train dataloader (in builders.py)
+            # then each epoch will repeat all your samples in the same order as the previous epoch, there is no shuffling
+            # to fix this, you should use MegatronPretrainingRandomBatchSampler instead, which alleviates this issue and allows
+            # random shuffling for each epoch.
+            raise ValueError(
+                "max_epochs > 1 is not supported unless using `MegatronPretrainingRandomBatchSampler` as the batch_sampler for your train dataloader"
+            )
+
+        self.model._reset_activation_checkpointing_args()
+        self.model._reset_sequence_parallelism_args()
+        set_eval(self.model)
+
+        if self.use_trtllm_generation:
+            self.trtllm_generate.refit(self.model)
+            clear_memory()
+
+        self.run_timer.start_time()
+
+        epoch_iter = range(self.epoch, self.cfg.max_epochs)
+        if len(epoch_iter) <= 0:
+            # epoch done
+            return
+
+        for _ in epoch_iter:
+            num_steps_in_epoch = min(
+                self.max_steps - self.step, self.num_steps_per_epoch - self.step % self.num_steps_per_epoch
+            )
+            loop_iter = range(num_steps_in_epoch)
+
+            if not loop_iter:
+                return  # training ended
+
+            global_pbar = tqdm(
+                self.augment_dataloader(self.train_dataloader),
+                initial=self.step,
+                total=self.max_steps,
+                leave=True,
+                desc="Generating steps",
+            )
+
+            for _, global_batch in zip(loop_iter, global_pbar):
+
+                dp_group = parallel_state.get_data_parallel_group()
+
+                gen_tokens = global_batch["prompt_and_response"]
+                prompt_lens = global_batch["prompt_lens"]
+                gen_lens = global_batch["gen_lens"]
+                valids = global_batch["valids"]
+
+                gen_tokens_list = [torch.zeros_like(gen_tokens) for _ in range(dp_group.size())]
+                prompt_lens_list = [torch.zeros_like(prompt_lens) for _ in range(dp_group.size())]
+                gen_lens_list = [torch.zeros_like(gen_lens) for _ in range(dp_group.size())]
+                valids_list = [torch.zeros_like(valids) for _ in range(dp_group.size())]
+
+                torch.distributed.all_gather(gen_tokens_list, gen_tokens, group=dp_group)
+                torch.distributed.all_gather(prompt_lens_list, prompt_lens, group=dp_group)
+                torch.distributed.all_gather(gen_lens_list, gen_lens, group=dp_group)
+                torch.distributed.all_gather(valids_list, valids, group=dp_group)
+
+                self.consumed_samples += self.model.cfg.global_batch_size
+                self.step += 1
+
+                if torch.distributed.get_rank() == 0 and gen_tokens_list is not None:
```

Comment:
As far as I can tell the second check is useless
```suggestion
                if torch.distributed.get_rank() == 0:
```